chia/full_node/coin_store.py

```diff
@@ -411,6 +411,83 @@ async def get_coin_states_by_ids(
 
         return coins
 
+    async def batch_coin_states_by_puzzle_hashes(
+        self,
+        puzzle_hashes: List[bytes32],
+        *,
+        min_height: uint32 = uint32(0),
+        include_spent: bool = True,
+        include_unspent: bool = True,
+        include_hinted: bool = True,
+        max_items: int = 50000,
+    ) -> Tuple[List[CoinState], uint32, bool]:
+        """
+        Returns the coin states, as well as the next block height and whether finished.
+        Note that the maximum number of puzzle hashes is currently set to 15000.
+        """
+
+        assert len(puzzle_hashes) <= 15000
+
+        coin_states: List[CoinState] = []
+
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            puzzle_hashes_db = tuple(puzzle_hashes)
+            puzzle_hash_count = len(puzzle_hashes_db)
+
+            if include_hinted:
+                cursor = await conn.execute(
+                    f"SELECT cr.confirmed_index, cr.spent_index, cr.coinbase, cr.puzzle_hash, "
+                    f"cr.coin_parent, cr.amount, cr.timestamp FROM coin_record cr "
+                    f"LEFT JOIN hints h ON cr.coin_name = h.coin_id "
+                    f'WHERE (cr.puzzle_hash in ({"?," * (puzzle_hash_count - 1)}?) '
+                    f'OR h.hint in ({"?," * (puzzle_hash_count - 1)}?)) '
+                    f"AND (cr.confirmed_index>=? OR cr.spent_index>=?) "
+                    f"{'' if include_spent else 'AND cr.spent_index=0'} "
+                    f"{'' if include_unspent else 'AND cr.spent_index>0'} "
```

Comment:
if both `include_spent` and `include_unspent` are `True`, you don't need either of these conditions. and I would expect sqlite to be a little bit happier without them

Reply:
I'm not sure what you mean? They aren't included in the query if both are True.

Reply:
Though if both are False (for some reason), it's probably better to return an empty list early to save time.

Reply:
what I mean is that if your SQL query includes `AND cr.spent_index=0 AND cr.spent_index>0`, it seems a bit redundant. But I see that I misread this. I think the `include_spent` and `unclude_inspent` don't quite mean what their names suggest. you would think that both being true would mean that both spent and unspent are included

Reply:
It's a bit confusing since it's backwards, but they do mean what they say. If `include_spent` is `False`, then you need to only select unspent coins. Otherwise you can include the spent ones too. And if `include_unspent` is `False`, then similarly you only select spent coins. If these are `True`, you don't need to add a filter since it's the default to include everything already. Hence the `'' if` (no additional filter). I think it works the way you want already - this is the same logic used above in other functions.